### ISSUE TITLE
TraceEvent scopes should be grouped by scope name only

### ISSUE THREAD

**Issue by uucidl** [jlfwong/speedscope#201]
I tested speedscope with TraceEvent files (generated with https://github.com/uucidl/uu.spdr)

When the scopes contain some arguments, speedscope helpfully shows it right there in the chart:

`Foo::Bar() {"Baz": 6}`

However it appears that aggregates/groups of scopes are using the scope name + args instead of
just the scope name. For the type of performance analysis I do, arguments are metadata around
the scope, but shouldn't be considered different scopes. 

(Note this was tested with the self-contained release of speedscope: speedscope-1.4.1)

**Comment by jlfwong** [comment]
Hi @uucidl!

Yeah, I figured that would be contentious when I was implementing it.

In the profiles I was testing on, like the ones generated by @aras_p's wonderful clang profiler (https://aras-p.info/blog/2019/01/16/time-trace-timeline-flame-chart-profiler-for-Clang/), metadata about individual files is stored in the details.

In the case of the profiles in that blogpost, if aggregation was done just on the basis of the name, then it would just say something like "80% is spent in `SourceFile`", which would be pretty unhelpful.

The option here that's most appealing to me is to continue with the default behavior that speedscope currently has, but provide a way of specifying in trace files that you don't want to include args in aggregration. This could be done by using the `{traceEvents: [...], ...}` format specified in the Trace Events spec. That might look something like this:

```
{
  traceEvents: [{...}, ...],
  speedscope: {
    ignoreArgs: true
  }
}
```

By far the easiest version of this would be to have it no longer display the args in the call frames at all then. Having call frames display different things in the time-ordered view but still be aggregated together in the left heavy and sandwich views isn't something speedscope supports at the moment. It's possible to build it, but it might be a gnarly data modeling change (I haven't thought too hard about it).

speedscope in general doesn't have any way to display arbitrary metadata at the moment, so the imports of those profiles would've been totally useless if that metadata wasn't included somehow.

Would providing a way to ignore `args` completely satisfy your use case? Another way of phrasing this is "if you ran a preprocessing script on your profile that stripped out all `args` before importing into speedscope, would that meet your needs"?

**Comment by uucidl** [comment]
I think I could live with the current behavior for now until speedscope eventually develops the ability to attach, visualize and analyze metadata on scopes in its other formats. I.e. if it includes them as a first class concept.

We're using uu.spdr as a performance trace format for commercial products, and for now the team uses chrome://tracing ; I'm on the lookout for other more advanced or more efficient visualisers, but this isn't that critical for now. It's already good to have speedscope as a backup plan, even if it requires some pre-processing.

**Comment by jlfwong** [comment]
Whoops, I used the wrong @ handle mention. @aras-p do you have opinions on how to handle this?

Part of the trickiness here is that (as far as I know) chrome://tracing doesn't do any kind of aggregation, so the spec has no need to specify under what conditions two events should be considered to be "the same" for aggregation.

Having aggregation be based on a subset of the keys in `args` seems okay to me, though it's perhaps a little bit uncomfortably magical. The traces from the clang profiler uses `{args: {detail: "<FILENAME HERE>"}}`, and having `"detail"` be one of the fields used for aggregation when others are not seems a bit odd to me.

**Comment by uucidl** [comment]
There is a form of aggregation in chrome:tracing: rubberband select a region with scopes, the check the lower pane: it shows average, total and self time grouped by scope name.

There’s even an histogram for cpu / clock time that appears when clicking one of these aggregations

**Comment by aras-p** [comment]
Yeah given that the "args" can be anything whatsoever (in one of our tools we put both "filename" and "time in milliseconds" -- since Chrome viewer displays in microseconds which can be hard to parse quickly)... I think aggregating only on the name would make sense as a "low effort" solution.

Having some sort of "optional aggregation" metadata would be cool too, but that sounds like more work :)

> By far the easiest version of this would be to have it no longer display the args in the call frames at all then

I, for one, love that currently it displays the args. Chrome tracing viewer does not, and that makes it much harder to see the context at a glance. We could put the filename into the title directly, but that would lose the "select all the compile steps" type of aggregation that Chrome viewer can do.

**Comment by jlfwong** [comment]
Doing aggregation only based on the name while simultaneously displaying args metadata would be pretty tricky in speedscope.

The clearest reason is that when viewing left-heavy view or sandwich view, you might aggregate two things with different args metadata, so which one would you display in the frame?

This also applies to the time-ordered view, since we combine stacks that are adjacent and identical.

I think my read from this thread is that doing aggregation only on name is a reasonable solution. To get the current behavior from a clang profiler output, we'd then need to pre-process to data to include the title directly.

This means that without preprocessing, args will not only not occur be visible in the flamegraph/flamecharts, they won't even be visible when you click into them.

Does that sound like an acceptable solution @aras-p and @uucidl?

**Comment by aras-p** [comment]
Well, dunno. Ideally it would have some sort of two modes:

1) maybe default, pay attention only to the event name.
2) also put in either everything, or some members of "args" into the name. Maybe via some optional fields in the json. E.g. I would use this to keep "filenames" in the args, so that it works with Chrome viewer nicely, but I'd emit something like `"speedscope_aggregate": "detail"` specially for speedscope.
